### PR TITLE
Provide JARs with OSGi metadata

### LINKED ISSUE #195: Provide JARs with OSGi metadata
The JARs that are produced during a build should contain OSGi metadata. This would allow to consume them as OSGi bundles right away.

### PR DESCRIPTION
**Pull request by rotty3000**
fixes  #195

Signed-off-by: Raymond Augé <raymond.auge@liferay.com>

## Overview

Provide JARs with OSGi metadata

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by sbrannen** [comment]
Tentatively slated for 5.6 M2 for _team discussion_

**Comment by sbrannen** [comment]
I'd really like to see some integration tests in place that actually verify that the generated OSGi metadata is consumable.

@junit-team/junit-committers, thoughts?

**Comment by rotty3000** [comment]
Great feedback. I wanted to make the first pass small concise. I'm happy to make any adjustments and I don't see any issues making the adjustments you want. Finally I can surely add a verification of the metadata. :)

**Comment by rotty3000** [comment]
> Where is the Export-Package statement?
fixed
> API guardian must not be imported
fixed
> Kotlin must always be an optional import.
wonderful
> Why is org.opentest4j the only import with version information?
good question! working on it.

**Comment by rotty3000** [comment]
made `Bundle-Name` === `${project.description}`

**Comment by sbrannen** [comment]
Did you forget to push the changes? Or... are you still working on the updates locally before pushing? (but.... absolutely no rush)

**Comment by rotty3000** [comment]
I've pushed all the above changes now. A major addition is only exporting packages marked with `@API` from all jars.

However there is an issue pending that I'm not sure why but for some reason bnd is not identifying the import versions for project dependencies. This is strange since I do see that bnd has the correct classpath and the jars are there. For example `junit-platform-engine` does not properly import `org.junit.platform.commons.*` with versions. I need to consult with @bjhargrave to figure out what might be the cause.

Also pending is verification of the metadata.

**Comment by rotty3000** [comment]
@sbrannen is it known that it's not possible to build any single module that has project dependencies? For instance when I try `./gradlew :junit-jupiter-engine:jar` it fails with
```
Execution failed for task ':junit-platform-commons:compileModule'.
> Extension with name 'sourceSets' does not exist. Currently registered extension names: [ext, eclipse, idea, defaultArtifacts, spotless]
```
it's not a huge deal, just makes debugging the build pretty hard if I can only execute the entire thing.

**Comment by sbrannen** [comment]
> `./gradlew :junit-jupiter-engine:jar`

I just executed that, and it worked for me.

Maybe you did a `clean` first?

**Comment by sbrannen** [comment]
Hmmm.... `./gradlew clean :junit-jupiter-engine:jar` works for me as well.

You executed that against `master` -- right?

**Comment by sbrannen** [comment]
> However there is an issue pending that I'm not sure why but for some reason bnd is not identifying the import versions for project dependencies. This is strange since I do see that bnd has the correct classpath and the jars are there. For example `junit-platform-engine` does not properly import `org.junit.platform.commons.*` with versions. I need to consult with @bjhargrave to figure out what might be the cause.
> 
> Also pending is verification of the metadata.

OK. Please keep us posted on both accounts.

File: build.gradle.kts [hunk 1]
```diff
@@ -140,6 +142,31 @@ subprojects {
...
+							!org.apiguardian.api,\
+							kotlin.*;resolution:="optional",\
+							*
+						-fixupmessages.kotlin.import: "Unused Import-Package instructions: \\[kotlin.*\\]";is:=ignore
+						-removeheaders: Private-Package
+						-noextraheaders: true
+
+						# The following exports only packages having classes annotated @API
+						# First collect all classes annotated with @API
+						possibleAPI: ${'$'}{classes;HIERARCHY_INDIRECTLY_ANNOTATED;org.apiguardian.api.API}
```
Comment (sbrannen) [inline build.gradle.kts]: Cool. I didn't know you could do things like that with `bnd`. 👍 

But... that will automatically export `org.junit.jupiter.engine.config` (which should not be exported because all classes annotated with `@API` in that package are actually `@API(status = INTERNAL, ...)`) -- right?

So it looks like we will still have to fine tune things a bit.

Or can you do even more magic with `bnd` to check the `status` attribute of `@API` declarations?

**Comment by rotty3000** [comment]
Let's see what we can come up with. ☺️ It might take me a bit. If you like
I can mark the PR as WIP. Or maybe you have a label...

File: settings.gradle.kts [hunk 2]
```diff
@@ -1,4 +1,12 @@
 pluginManagement {
+	resolutionStrategy {
+		eachPlugin {
+			if (requested.id.id == "biz.aQute.bnd.builder") {
+				useModule("biz.aQute.bnd:biz.aQute.bnd.gradle:4.3.0")
+			}
+		}
+	}
+
```
Comment (marcphilipp) [inline settings.gradle.kts]: Please add the version to the `plugins` block instead and delete `resolutionStrategy` again.

File: settings.gradle.kts [hunk 3]
```diff
@@ -10,6 +18,12 @@ pluginManagement {
 		id("org.asciidoctor.convert") version "1.5.8.1"
 		id("me.champeau.gradle.jmh") version "0.4.8"
 		id("io.spring.nohttp") version "0.0.3.RELEASE"
+		id("biz.aQute.bnd.builder")
+	}
+
+	repositories {
+		gradlePluginPortal()
+		mavenCentral()
```
Comment (marcphilipp) [inline settings.gradle.kts]: Why do we need `mavenCentral()`?

File: build.gradle.kts [hunk 4]
```diff
@@ -140,6 +142,31 @@ subprojects {
 		}
 
 		afterEvaluate {
+			if (project != project(":junit-jupiter")) {
+				tasks["jar"].withConvention(aQute.bnd.gradle.BundleTaskConvention::class) {
+					setBnd("""
```
Comment (marcphilipp) [inline build.gradle.kts]: Doesn't `bnd = """` work?

File: build.gradle.kts [hunk 5]
```diff
@@ -140,6 +142,31 @@ subprojects {
 		}
 
 		afterEvaluate {
+			if (project != project(":junit-jupiter")) {
+				tasks["jar"].withConvention(aQute.bnd.gradle.BundleTaskConvention::class) {
```
Comment (marcphilipp) [inline build.gradle.kts]: Will this work for subprojects that have disabled the `jar` task and use `shadowJar` instead?

File: build.gradle.kts [hunk 6]
```diff
@@ -140,6 +142,31 @@ subprojects {
 		}
 
 		afterEvaluate {
+			if (project != project(":junit-jupiter")) {
```
Comment (marcphilipp) [inline build.gradle.kts]: Why is this in an `afterEvaluate` block?

Comment (marcphilipp) [inline build.gradle.kts]: Please add this to `java-library-conventions.gradle.kts` instead of to the root build script.

File: settings.gradle.kts (hunk 2 above)
Comment (rotty3000) [reply settings.gradle.kts]: Due to an issue with the gradle plugin portal reported here https://github.com/gradle/plugin-portal-requests/issues/9 version 4.3.0 is not accessible and so we have to use the approach shown until the issue can be resolved. Sorry about that.

File: settings.gradle.kts (hunk 3 above)
Comment (rotty3000) [reply settings.gradle.kts]: Also due to the issue with the gradle plugin portal...

File: build.gradle.kts (hunk 4 above)
Comment (rotty3000) [reply build.gradle.kts]:
```
Script compilation error:

  Line 147:                                     bnd = """
                 ^ Val cannot be reassigned
```

File: build.gradle.kts (hunk 5 above)
Comment (rotty3000) [reply build.gradle.kts]: That's a good question. I only briefly looked at `shadowJar` but I wasn't clear it's real purpose. Now that you mention it I'll have to dig deeper into it. If you could fill me in with a small description of what the intention is there it'll help me make the best choice on how to handle it.

File: build.gradle.kts (hunk 6 above)
Comment (rotty3000) [reply build.gradle.kts]: Because that was the only place I could get  `${project.description}` to resolve correctly. Outside of this I was getting null. Any suggestions to work around that would be appreciated.

Comment (rotty3000) [reply build.gradle.kts]: Great, I wasn't aware of it's existence.

Comment (rotty3000) [reply build.gradle.kts]: I managed to move the bnd configuration over to `java-library-conventions.gradle.kts`

Comment (rotty3000) [reply build.gradle.kts]: this was also solved.

File: build.gradle.kts (hunk 5 above)
Comment (marcphilipp) [reply build.gradle.kts]: We use the shadow plugin to build a fat jar with relocated dependencies, e.g. in `junit-jupiter-params`. You could probably use the (not very pretty) approach from here to catch both tasks: https://github.com/junit-team/junit5/blob/master/buildSrc/src/main/kotlin/java-library-conventions.gradle.kts#L118

File: settings.gradle.kts (hunk 2 above)
Comment (marcphilipp) [reply settings.gradle.kts]: I see. Thanks for the explanation!

File: buildSrc/src/main/kotlin/java-library-conventions.gradle.kts [hunk 7]
```diff
@@ -180,6 +183,29 @@ tasks.jar {
 				"Implementation-Vendor" to "junit.org"
 		)
 	}
+
+	withConvention(BundleTaskConvention::class) {
+		setBnd("""
+			Bundle-Name: ${project.description}
```
Comment (marcphilipp) [inline buildSrc/src/main/kotlin/java-library-conventions.gradle.kts]: Isn't this evaluated before the descriptions are set?

File: buildSrc/src/main/kotlin/java-library-conventions.gradle.kts [hunk 8]
```diff
@@ -180,6 +183,29 @@ tasks.jar {
...
+				*
+			-fixupmessages.kotlin.import: "Unused Import-Package instructions: \\[kotlin.*\\]";is:=ignore
+			-removeheaders: Private-Package
+			-noextraheaders: true
+
+			# The following exports only packages having classes annotated @API
+			# First collect all classes annotated with @API
+			possibleAPI: ${'$'}{classes;HIERARCHY_INDIRECTLY_ANNOTATED;org.apiguardian.api.API}
+			# Strip off the classNames to obtain a list of package names
+			possiblePackages: ${'$'}{replace;${'$'}{possibleAPI};\.[^.]*${'$'};}
```
Comment (marcphilipp) [inline buildSrc/src/main/kotlin/java-library-conventions.gradle.kts]: I'm curious if we could we parse the module descriptors instead... 🤔

Comment (sbrannen) [reply buildSrc/src/main/kotlin/java-library-conventions.gradle.kts]: That would be great it we could!

Better to not duplicate this information.

Comment (sormuras) [reply buildSrc/src/main/kotlin/java-library-conventions.gradle.kts]: We could parse their source code (similar to [this](https://github.com/sormuras/bach/blob/master/src/bach/Bach.java#L697-L728)) or use the already compiled descriptors -- they are compiled at this stage, aren't they?

**Comment by kriegfrj** [comment]
I like where this project is going, thanks guys. A couple of comments:

* Given the way that JUnit 5 is versioned, perhaps we could consider making use of OSGi contracts instead of straight-out import/export package mechanism? I don't have a lot of first-hand experience with OSGi contracts, but from what little knowledge I do have, they might be a fit. Perhaps @bjhargrave can comment on this.

* I would also like to see us leverage the OSGi capability mechanism in some way to define dependency relationships between test bundles and test engine bundles. That way, the resolver can help with making sure that the correct engines are included in the set of run bundles. For example, on the test engine bundle side, something like this:

```
Provide-Capability: org.junit.platform.engine;
    org.junit.platform.engine.id='junit-jupiter;version="[5.5.1]"' # By convention, make this == TestEngine.getId()
    version=5.5.1
```

...then on the test bundle side, you can declare (eg) the following:

```
Require-Capability: org.junit.platform.engine;
   org.junit.platform.engine.id='junit-jupiter;version="[5.5,6)",junit-vintage;version=4.12,org.myorg.my-test-engine;version=1.0'
```

Now the resolver will be able to ensure that the test bundles are paired with the correct versions of the correct test engines. The syntax might need a bit of tweaking (I'm not an expert in this area), but hopefully the idea makes sense.

File: build.gradle.kts (hunk 5 above)
Comment (rotty3000) [reply build.gradle.kts]: k, this is solved.

File: buildSrc/src/main/kotlin/java-library-conventions.gradle.kts (hunk 7 above)
Comment (rotty3000) [reply buildSrc/src/main/kotlin/java-library-conventions.gradle.kts]: I split this up so that only that one pieces is during `afterEvaluate`

**Comment by rotty3000** [comment]
Ok, only pending are:
- the `@API(status = INTERNAL, ...)`
- the verification

File: buildSrc/src/main/kotlin/java-library-conventions.gradle.kts (hunk 8 above)
Comment (rotty3000) [reply buildSrc/src/main/kotlin/java-library-conventions.gradle.kts]: The module descriptor will always trail your actual code. You should let tooling give you a result _as close to perfect as a algorithm can bring it_ and then let you take it the last mile. This will eliminate guess work and automatically find details humans aren't good at identifying.

In short, let's leave that part for a future update, shall we? :)

Comment (rotty3000) [reply buildSrc/src/main/kotlin/java-library-conventions.gradle.kts]: This is not to say that I will not solve the @API exports issue... but rather then parsing of the module descriptor suggestion. ;)

**Comment by rotty3000** [comment]
@kriegfrj let's not shock our goodly junit developers with too much too early :D 

However, I do like your suggestion. Maybe you could open a separate issue specific to capabilities and requirements?

File: buildSrc/src/main/kotlin/java-library-conventions.gradle.kts (hunk 8 above)
Comment (rotty3000) [reply buildSrc/src/main/kotlin/java-library-conventions.gradle.kts]: ... if that's ok with you all!

**Comment by kriegfrj** [comment]
> @kriegfrj let's not shock our goodly junit developers with too much too early :D
>

:smile: sorry!

> However, I do like your suggestion. Maybe you could open a separate issue specific to capabilities and requirements?

Sure - good idea to keep it separate. Where should I raise the issue - here, on bnd github, or both?

**Comment by rotty3000** [comment]
I'd suggest on this project since that's where you intended to make the change, IIUC.

File: buildSrc/src/main/kotlin/java-library-conventions.gradle.kts (hunk 8 above)
Comment (marcphilipp) [reply buildSrc/src/main/kotlin/java-library-conventions.gradle.kts]: Fair enough, we could instead add an integration test that verifies the two things are aligned.

**Comment by rotty3000** [comment]
Ok, so, the `@API(status = INTERNAL)` is solved. That also allowed me to do a big cleanup. the changes are much more concise again. You'll note that I added an _inline_ bnd plugin benefiting from the kotling buildSrc that was already there. But do note that bnd is adding support for APIGuardian in the future release (but I didn't want to wait so I added the plugin here).

Only verification remains.

File: buildSrc/src/main/kotlin/java-library-conventions.gradle.kts (hunk 8 above)
Comment (rotty3000) [reply buildSrc/src/main/kotlin/java-library-conventions.gradle.kts]: resolving this since we have a solution now

**Comment by rotty3000** [comment]
Ok, so before we move on, I wanted to point out that all this did find some oddities:

- `org.junit.platform.commons.logging` is not exported because all it's types are marked as `@API(status = INTERLAN)` however it's imported by almost all other modules.

So, what we're thinking is that _really_ `@API` should always be exported, but that the package exports should also always be marked with their `status` attribute. In the case of INTERNAL, it's really an API for junit itself, like a `exports ... to ...` in JPMS. We can do that also in OSGi. So I'm going to change the heuristic.

**Comment by sbrannen** [comment]
> Ok, so, the `@API(status = INTERNAL)` is solved. 

Great

> But do note that bnd is adding support for APIGuardian in the future release

Oh wow. That's a pleasant surprise.

> Only verification remains.

👍

**Comment by rotty3000** [comment]
Ok, so this PR is finally feature complete, including verification.

File: gradle.properties [hunk 9]
```diff
@@ -17,3 +17,8 @@ org.gradle.jvmargs=-Xmx1g -XX:MaxMetaspaceSize=512m -XX:+HeapDumpOnOutOfMemoryEr
 org.gradle.caching=true
 org.gradle.parallel=true
 
+# Use this flag so that when compiling against project dependencies
+# the jar is created and used instead of using src/main/classes.
+#
+# See https://docs.gradle.org/current/userguide/java_library_plugin.html#sub:java_library_known_issues_windows_performance
+systemProp.org.gradle.java.compile-classpath-packaging=true
```
Comment (marcphilipp) [inline gradle.properties]: Why is this required?

**Comment by rotty3000** [comment]
@marcphilipp I will add some comments around all the tasks to explain their use. Standby! :)

File: gradle.properties (hunk 9 above)
Comment (rotty3000) [reply gradle.properties]: Ok, let me see. 

It begins with this very good question by @sbrannen:
> Why is org.opentest4j the only import with version information?

This was also a surprise to me because bnd gradle plugin was normally always able to determine the version to import a package because it would scan the classpath (essentially `project.sourceSet.main.compile`) and observe the package versions it found within the __jars__ that are there.

Remember, the bnd gradle plugin only adds metadata to the __jar__.

However, this was not working on this project! Why not?

I observed that this project uses the `java-library` gradle plugin which is an observably uncommon scenario as most gradle projects use the `java` plugin.

The distinctive difference here is that the `java` gradle plugin having dependencies of type `project` (e.g. `compile project(":foo")`) will result the project __jars__ being in `project.sourceSet.main.compile`. As I indicated earlier, this allowed bnd to observe the OSGi metadata for those dependent projects.

Now, the `java-library` plugin does _not_ by default cause a project dependency's jar to be built, AND when looking at the ``project.sourceSet.main.compile` will only contain the project output directory. Bnd does not emit metadata into the output directory and therefore there is no metadata to observer.

Setting the flag above to `true` tells the `java-library` plugin to revert to building the jar and using this in `project.sourceSet.main.compile`.

Comment (rotty3000) [reply gradle.properties]: Also, as I indicated in the comment, you can read more about this here https://docs.gradle.org/current/userguide/java_library_plugin.html#sub:java_library_known_issues_windows_performance

Comment (rotty3000) [reply gradle.properties]: Or better https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_classes_usage

**Comment by rotty3000** [comment]
@marcphilipp I added lots of comments, but please if you have more questions I can take those too.

File: buildSrc/build.gradle.kts [hunk 10]
```diff
@@ -10,4 +10,5 @@ repositories {
 dependencies {
 	implementation(kotlin("gradle-plugin"))
 	implementation("de.marcphilipp.gradle:nexus-publish-plugin:0.4.0")
+	implementation("biz.aQute.bnd:biz.aQute.bnd.gradle:4.3.0")
```
Comment (kriegfrj) [inline buildSrc/build.gradle.kts]: Should we bump to 4.3.1 now that the maintenance release is publicly available?

File: junit-platform-console-standalone/junit-platform-console-standalone.gradle.kts [hunk 11]
```diff
@@ -59,4 +74,10 @@ tasks {
 			))
 		}
 	}
+
+	// This jar contains some Java 9 code (I think from shadowed jansi)
+	// So in order to resolve this, it can only run on Java 9
+	verifyOSGiProperties {
+		property("-runee", "JavaSE-9")
+	}
```
Comment (marcphilipp) [inline junit-platform-console-standalone/junit-platform-console-standalone.gradle.kts]: This feels wrong to me. Do you have a failure message?

File: buildSrc/src/main/kotlin/java-library-conventions.gradle.kts [hunk 12]
```diff
@@ -139,6 +143,105 @@ if (project in mavenizedProjects) {
...
+				org.junit.platform.commons.logging;status=INTERNAL,\
+				kotlin.*;resolution:="optional",\
+				*
+
+			# This tells bnd not to complain if a module doesn't actually import
+			# the kotlin packages, but enough modules do to make it a default.
+			-fixupmessages.kotlin.import: "Unused Import-Package instructions: \\[kotlin.*\\]";is:=ignore
+
+			# This tells bnd to ignore classes it files in `META-INF/versions/`
+			# because bnd doesn't yet support multi-version jars.
```
Comment (marcphilipp) [inline buildSrc/src/main/kotlin/java-library-conventions.gradle.kts]:
```suggestion
			# because bnd doesn't yet support multi-release jars.
```

Comment (rotty3000) [reply buildSrc/src/main/kotlin/java-library-conventions.gradle.kts]: fixing

File: junit-platform-console-standalone/junit-platform-console-standalone.gradle.kts (hunk 11 above)
Comment (rotty3000) [reply junit-platform-console-standalone/junit-platform-console-standalone.gradle.kts]: I will check to see exactly what classes are compiled with Java 9 bytecode (bnd uses the class' magic number to make this determination).

Comment (rotty3000) [reply junit-platform-console-standalone/junit-platform-console-standalone.gradle.kts]: The class in question is `org.junit.platform.console.ConsoleLauncherToolProvider` which implements `java.util.spi.ToolProvider` which is `@since 9` :)

**Comment by rotty3000** [comment]
As requested I added a few checks to the manifest test.

**Comment by rotty3000** [comment]
failures here do not seem to be related. :(

**Comment by rotty3000** [comment]
See the same failures on this other sanity check PR https://github.com/junit-team/junit5/pull/2101

**Comment by sormuras** [comment]
`Manifest attribute Bundle-Version expected to be 5.6.0-SNAPSHOT, but is: 5.6.0.SNAPSHOT`

There's `.` written to the manifest instead of a `-`. Who generates the string value that's used for `Bundle-Version`?

**Comment by rotty3000** [comment]
Ah, I didn't see this error. Dot is required by OSGi version syntax. I will parse the version into the correct syntax for this check. Standby!
